keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+other HC endpoints in use today, such as: `NodePort` or the Kubelet's healthz
+port). Besides allowing us to document it better and provide the user with a
+better user experience in general, we would also avoiding mixing different
+features depending on specific fields (ex: `NodePort` - which serves as a
+back-bone to a lot of features) and use it as a "dedicated HC intended
+attribute". Given that we would extend it, we also would not need to worry about
+backport/version skew compatibility - which would have be the case should we opt
+for another mechanism which might impact existing class B functionality
+dependent upon `healthCheckNodePort`.
+
+Clearly a big difficulty with doing this will be to make sure that whatever is
+configured for class B services, remains to only be configured for class B
+services - even though both classes of services will have this field.
+
+Doing this would require: changing the API for class A services, change the HCs
+configured by cloud providers to point to the `healthCheckNodePort` for all
+services, and change the service proxies to open ports for all services which
+have this field, but respond differently depending on if the service is class A
+or B. If A: check healthz, if B: check healthz and if there's a local endpoint.
+
+### User Stories (Optional)
+
+#### "Kube version N after KEP" - Extend `healthCheckNodePort` to ETP=cluster
+
+We need to being by performing the necessary changes to have class A services
+get assigned a `healthCheckNodePort`. This means, besides the implementation,
+that we validate that no component in Kubernetes acts solely on this field to
+perform setup for services which are class B. If that is the case: we need to
+change those components too.
+
+#### "Kube version N after KEP" - Implement service proxy logic
+
+Once the API for class A services has changed to include them getting
+`healthCheckNodePort` assigned, we need to update Kube-proxy with the logic
+which will, a) open ports for class A and B services, b) return a different
+response depending on if the HC is intended for a class A or B service. The
+logic to have the health check server for class B services validate Kube-proxy's
+healthz is already being implemented in 
+[Kubernetes PR #111661](https://github.com/kubernetes/kubernetes/pull/111661),
+what might remain is: stop having Kube-proxy's healthz server be bound to the
+node IP addresses (even be a listening HTTP server) - depending on if other
+clients is using it or not.
+
+#### "Kube version N + X after KEP" - Point HCs to target `healthCheckNodePort` for all services
+
+Due to the version skew that Kubernetes supports between the control and data
+plane we would need to wait X releases until we actually migrate the configured
+HC for all class A type services to point to the `healthCheckNodePort`. Once
+that is done however, cloud provider implementations will need to point all HCs
```

Comment:
that is the part that I'm skeptical, I have doubts cloud providers will stop doing this, but I'll be happy to proof wrong

Reply:
Me too.  I think we can change kube-proxy HC to answer "is this node OK to use in the data-plane" as we think is best, and we can suggest that impls should use it, but I am not sure we can FORCE them to.

Reply:
Well, you can't enforce external providers (either service proxies, cloud providers, CNI plugins or other) to do anything...but even support features. But we can give guidelines of what we think is best and propose improvements that they can follow. Customers/users will force them to align :)